portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -15,18 +15,93 @@
 package com.google.sps.servlets;
 
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.lang.Integer;
+
+import com.google.sps.data.Comment;
+import com.google.gson.Gson;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
-@WebServlet("/data")
+/** Servlet that returns comments on portfolio */
+@WebServlet("/comments")
 public class DataServlet extends HttpServlet {
+  private int commentLimit = 25; //default set to 25
+  private final static String contentType = "application/json;charset=utf-8";
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello from Antonia!</h1>");
+    String parameterName = "number";
+    String number = request.getParameter(parameterName);
+
+    if (number != null) {
+      this.commentLimit = Integer.parseInt(number);
+    }
+
+    response.setContentType(contentType);
+
+    List<Comment> comments = new ArrayList<>();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Query query = new Query("Comment").addSort("date", SortDirection.DESCENDING);
+
+    PreparedQuery preparedQuery = datastore.prepare(query);
+    List<Entity> results = preparedQuery.asList(FetchOptions.Builder.withLimit(this.commentLimit));
+    
+    for (Entity entity : results) {
+      String entityName = (String) entity.getProperty("name");
+      String entityText = (String) entity.getProperty("text");
+      Long entityID = (Long) entity.getProperty("id");
+      Date date = (Date) entity.getProperty("date");
+
+      Comment databaseComment = new Comment(entityName, entityText, entityID, date);
+
+      comments.add(databaseComment);
+    }
+    Gson myGson = new Gson();
+    String json = myGson.toJson(comments);
+
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType(contentType);
+    String name = getParameter(request, "name", "Anonymous");
+    String comment = getParameter(request, "comment", "");
+    Date createDate = new Date();
+
+    if (comment == "" || comment == null) {
+      response.sendRedirect("/");
```

Comment:
It's a good style to return error messages rather than having a behavior identitcal to the correct case.

OTOH, this case would be hit only if someone bypasses the checks done in Javascript.

BTW, yes, the Javascript checks can be bypassed, so if someone plays around with your page, they can still inject "nonconformant" data. Given this is only a "toy" project, I think is fine.